lfs/gitscanner_catfilebatchcheck.go

```diff
@@ -19,99 +18,87 @@ func runCatFileBatchCheck(smallRevCh chan string, revs *StringChannelWrapper, er
 		return err
 	}
 
-	go catFileBatchCheckOutput(smallRevCh, cmd, errCh)
-	go catFileBatchCheckInput(cmd, revs, errCh)
-	return nil
-}
-
-func catFileBatchCheckOutput(smallRevCh chan string, cmd *wrappedCmd, errCh chan error) {
-	scanner := &catFileBatchCheckScanner{s: bufio.NewScanner(cmd.Stdout)}
-	for scanner.Scan() {
-		smallRevCh <- scanner.BlobOID()
-	}
+	go func() {
+		scanner := &catFileBatchCheckScanner{s: bufio.NewScanner(cmd.Stdout), limit: blobSizeCutoff}
+		for r := range revs.Results {
+			cmd.Stdin.Write([]byte(r + "\n"))
+			hasNext := scanner.Scan()
+			if b := scanner.BlobOID(); len(b) > 0 {
+				smallRevCh <- b
+			}
+
+			if err := scanner.Err(); err != nil {
+				errCh <- err
+			}
+
+			if !hasNext {
+				break
+			}
+		}
 
-	if err := scanner.Err(); err != nil {
-		errCh <- err
-	}
+		if err := revs.Wait(); err != nil {
+			errCh <- err
+		}
+		cmd.Stdin.Close()
 
-	stderr, _ := ioutil.ReadAll(cmd.Stderr)
-	err := cmd.Wait()
-	if err != nil {
-		errCh <- fmt.Errorf("Error in git cat-file --batch-check: %v %v", err, string(stderr))
-	}
-	close(smallRevCh)
-	close(errCh)
-}
+		stderr, _ := ioutil.ReadAll(cmd.Stderr)
+		err := cmd.Wait()
+		if err != nil {
+			errCh <- fmt.Errorf("Error in git cat-file --batch-check: %v %v", err, string(stderr))
+		}
+		close(smallRevCh)
+		close(errCh)
+	}()
 
-func catFileBatchCheckInput(cmd *wrappedCmd, revs *StringChannelWrapper, errCh chan error) {
-	for r := range revs.Results {
-		cmd.Stdin.Write([]byte(r + "\n"))
-	}
-	err := revs.Wait()
-	if err != nil {
-		// We can share errchan with other goroutine since that won't close it
-		// until we close the stdin below
-		errCh <- err
-	}
-	cmd.Stdin.Close()
+	return nil
 }
 
 type catFileBatchCheckScanner struct {
 	s       *bufio.Scanner
+	limit   int
 	blobOID string
-	err     error
 }
 
 func (s *catFileBatchCheckScanner) BlobOID() string {
 	return s.blobOID
 }
 
 func (s *catFileBatchCheckScanner) Err() error {
-	return s.err
+	return s.s.Err()
```

Comment:
Yeah, I could go either way here. I just went with the one with fewer LOC. I have a hunch that if this code is refactored to get rid of `*bufio.Scanner`, we'll be moving to libgit2 or something completely different.